flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,340 @@
+	private final AtomicLong currentMaxAttemptId = new AtomicLong(0);
+
+	/** Current max pod index. When creating a new pod, it should increase one. */
+	private final AtomicLong currentMaxPodId = new AtomicLong(0);
+
+	private final String clusterId;
+
+	private final FlinkKubeClient kubeClient;
+
+	private final ContaineredTaskManagerParameters taskManagerParameters;
+
+	private final List<String> taskManagerStartCommand;
+
+	/** The number of pods requested, but not yet granted. */
+	private int numPendingPodRequests = 0;
+
+	public KubernetesResourceManager(
+			RpcService rpcService,
+			String resourceManagerEndpointId,
+			ResourceID resourceId,
+			Configuration flinkConfig,
+			HighAvailabilityServices highAvailabilityServices,
+			HeartbeatServices heartbeatServices,
+			SlotManager slotManager,
+			JobLeaderIdService jobLeaderIdService,
+			ClusterInformation clusterInformation,
+			FatalErrorHandler fatalErrorHandler,
+			ResourceManagerMetricGroup resourceManagerMetricGroup) {
+		super(
+			flinkConfig,
+			System.getenv(),
+			rpcService,
+			resourceManagerEndpointId,
+			resourceId,
+			highAvailabilityServices,
+			heartbeatServices,
+			slotManager,
+			jobLeaderIdService,
+			clusterInformation,
+			fatalErrorHandler,
+			resourceManagerMetricGroup);
+		this.clusterId = flinkConfig.getString(KubernetesConfigOptions.CLUSTER_ID);
+		this.defaultCpus = flinkConfig.getDouble(KubernetesConfigOptions.TASK_MANAGER_CPU, numSlotsPerTaskManager);
+
+		this.kubeClient = createFlinkKubeClient();
+
+		this.taskManagerParameters =
+			ContaineredTaskManagerParameters.create(flinkConfig, taskExecutorResourceSpec, numSlotsPerTaskManager);
+
+		this.taskManagerStartCommand = getTaskManagerStartCommand();
+	}
+
+	@Override
+	protected Configuration loadClientConfiguration() {
+		return GlobalConfiguration.loadConfiguration();
+	}
+
+		@Override
+	protected void initialize() throws ResourceManagerException {
+		recoverWorkerNodesFromPreviousAttempts();
+
+		kubeClient.watchPodsAndDoCallback(getTaskManagerLabels(), this);
+	}
+
+	@Override
+	public CompletableFuture<Void> onStop() {
+		// shut down all components
+		Throwable exception = null;
+
+		try {
+			kubeClient.close();
+		} catch (Throwable t) {
+			exception = t;
+		}
+
+		return getStopTerminationFutureOrCompletedExceptionally(exception);
+	}
+
+	@Override
+	protected void internalDeregisterApplication(ApplicationStatus finalStatus, @Nullable String diagnostics) {
+		LOG.info(
+			"Stopping kubernetes cluster, clusterId: {}, diagnostics: {}",
+			clusterId,
+			diagnostics == null ? "" : diagnostics);
+		kubeClient.stopAndCleanupCluster(clusterId);
+	}
+
+	@Override
+	public Collection<ResourceProfile> startNewWorker(ResourceProfile resourceProfile) {
+		LOG.info("Starting new worker with resource profile, {}", resourceProfile);
+		if (!resourceProfilesPerWorker.iterator().next().isMatching(resourceProfile)) {
+			return Collections.emptyList();
+		}
+		requestKubernetesPod();
```

Comment:
I am not quite sure about how to guarantee the new requested pod can satisfy the required resource, because we did  not consider the required `resourceProfile` while `requestKubernetesPod`?

Reply:
Currently the `resourceProfile` is UNKNOWN for data stream job by default and with managed memory set for batch job. If the required `resourceProfile` could not match the `resourceProfilesPerWorker`, then it will never be allocated. We start a new taskmanager pod only when it matches.